java-checks/src/main/java/org/sonar/java/checks/BoxedBooleanExpressionsCheck.java

```diff
@@ -0,0 +1,119 @@
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.DoWhileStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+
+@Rule(key = "S5411")
+public class BoxedBooleanExpressionsCheck extends IssuableSubscriptionVisitor {
+
+  private static final String BOOLEAN = "java.lang.Boolean";
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return Arrays.asList(Kind.FOR_STATEMENT, Kind.WHILE_STATEMENT, Kind.DO_STATEMENT, Kind.IF_STATEMENT, Kind.EQUAL_TO,
+        Kind.NOT_EQUAL_TO, Kind.CONDITIONAL_EXPRESSION, Kind.CONDITIONAL_AND, Kind.CONDITIONAL_OR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    ExpressionTree expression;
+    switch (tree.kind()) {
+      case FOR_STATEMENT:
+        expression = ((ForStatementTree) tree).condition();
+        break;
+      case WHILE_STATEMENT:
+        expression = ((WhileStatementTree) tree).condition();
+        break;
+      case DO_STATEMENT:
+        expression = ((DoWhileStatementTree) tree).condition();
+        break;
+      case IF_STATEMENT:
+        expression = ((IfStatementTree) tree).condition();
+        break;
+      case CONDITIONAL_EXPRESSION:
+        expression = ((ConditionalExpressionTree) tree).condition();
+        break;
+      case EQUAL_TO:
+      case NOT_EQUAL_TO:
+      case CONDITIONAL_AND:
+      case CONDITIONAL_OR:
+        expression = (BinaryExpressionTree) tree;
+        break;
+      default:
+        expression = null;
+    }
+    if (expression != null) {
+      ExpressionTree boxed = findBoxedBoolean(expression);
+      if (boxed != null) {
+        reportIssue(boxed, "Use the primitive boolean expression here.");
+      }
+    }
+  }
+
+  @CheckForNull
+  private static ExpressionTree findBoxedBoolean(ExpressionTree tree) {
+    if (isBooleanIdentifier(tree)) {
+      return tree;
+    }
+    if (tree instanceof BinaryExpressionTree) {
+      BinaryExpressionTree expression = (BinaryExpressionTree) tree;
+      if (!isNullCheck(expression) && !isNullCheck(expression.leftOperand())) {
+        if (isBooleanIdentifier(expression.leftOperand())) {
+          return expression.leftOperand();
+        }
+        if (isBooleanIdentifier(expression.rightOperand())) {
+          return expression.rightOperand();
+        }
+      }
+    }
+    return null;
+  }
+
+  private static boolean isBooleanIdentifier(ExpressionTree tree) {
+    return tree.symbolType().is(BOOLEAN) && tree.is(Kind.IDENTIFIER);
```

Comment:
It's sufficient to test for `java.lang.Boolean` type, that way we can also support usage of functions with `java.lang.Boolean` return value (don't forget to add test)